From 479781da278d61f8813a4f1b9159d7d4cb76c9df Mon Sep 17 00:00:00 2001
From: "Doug.Smith" <doug.smith@lairdconnect.com>
Date: Wed, 8 Jun 2022 16:13:55 -0400
Subject: [PATCH] PKCS12: work-around for FIPS pkcs12 failure

FIPS mode cannot load a pkcs12 file.
MAC verification fails due to openssl-3.0 fips blocking
use of the PKCS12KDF as a non-fips approved KDF.
(Although fips 140-2 implementation guidance
suggests that non-approved is allowed
for pkcs12...)

This work-around allows pkcs12 without the optional mac.
To generate a pkcs12 file with no mac
Use: openssl pksc12 -nomac

TBD: allow PKCS12KDF in fips mode
---
 crypto/pkcs12/p12_kiss.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/crypto/pkcs12/p12_kiss.c b/crypto/pkcs12/p12_kiss.c
index 229b34c..15379d4 100644
--- a/crypto/pkcs12/p12_kiss.c
+++ b/crypto/pkcs12/p12_kiss.c
@@ -67,6 +67,12 @@ int PKCS12_parse(PKCS12 *p12, const char *pass, EVP_PKEY **pkey, X509 **cert,
             ERR_raise(ERR_LIB_PKCS12, PKCS12_R_MAC_VERIFY_FAILURE);
             goto err;
         }
+#if 1
+    } else if (!PKCS12_mac_present(p12)) {
+        /* optional mac not present -- skip verification */
+        /* work-around for PKCS12 mac validation failure in FIPS mode */
+        ;
+#endif
     } else if (!PKCS12_verify_mac(p12, pass, -1)) {
         ERR_raise(ERR_LIB_PKCS12, PKCS12_R_MAC_VERIFY_FAILURE);
         goto err;
-- 
2.17.1

